server/server/src/billing/billing_service.rs

```diff
@@ -379,113 +514,300 @@ pub async fn stripe_webhooks(
+        if let NotificationType::SubscriptionPurchased = notification_type {
+            return Ok(());
+        }
+
+        let account_id = &subscription
+            .obfuscated_external_account_id
+            .clone()
+            .ok_or_else(|| {
+                internal!("There should be an account id attached to a purchase: {:?}", sub_notif)
+            })?;
+
+        info!("Retrieved full subscription info for notification event {:?} with an obfuscated id of {:?}", notification_type, account_id);
+
+        let public_key: PublicKey = con
+            .maybe_json_get(keys::public_key_from_gp_account_id(account_id))
+            .await?
+            .ok_or_else(|| {
+                internal!("There is no public_key related to this account_id: {:?}", account_id)
+            })?;
+
+        info!("Updating subscription history to match new subscription state.");
+
+        let tx_result = tx!(&mut con, pipe, &[keys::subscription_profile(&public_key)], {
+            match lock_billing_workflow(
+                &public_key,
+                &mut con,
+                server_state.config.stripe.millis_between_user_payment_flows,
             )
-        })?;
+            .await
+            {
+                Ok(mut sub_profile) => {
+                    if let Some(BillingPlatform::GooglePlay(ref mut info)) =
+                        sub_profile.billing_platform
+                    {
+                        match notification_type {
+                            NotificationType::SubscriptionRecovered => {
+                                info.account_state = GooglePlayAccountState::Ok;
+                                info.expiration_time = subscription.borrow().expiry_time_millis.as_ref().ok_or_else(|| Abort(internal!("Cannot get expiration time of a recovered subscription. public_key {:?}, subscription notification type: {:?}", public_key, notification_type)))?.parse().map_err(|e| Abort(internal!("Cannot parse millis into int: {:?}", e)))?;
+                                sub_profile.data_cap = PREMIUM_TIER_USAGE_SIZE;
+
+                                sub_profile.last_in_payment_flow = 0;
+                                pipe.json_set(keys::subscription_profile(&public_key), &sub_profile)
+                                // give back premium and update time exp
+                            }
+                            NotificationType::SubscriptionRenewed
+                            | NotificationType::SubscriptionRestarted => {
+                                info.account_state = GooglePlayAccountState::Ok;
+                                info.expiration_time = subscription.borrow().expiry_time_millis.as_ref().ok_or_else(|| Abort(internal!("Cannot get expiration time of a recovered subscription. public_key {:?}, subscription notification type: {:?}", public_key, notification_type)))?.parse().map_err(|e| Abort(internal!("Cannot parse millis into int: {:?}", e)))?;
```

Comment:
you should newline separate all these. cargo fmt just gives up if a lines too long, it's pretty stupid. same for all below. Also they look a lil copy pasty to me... but I'll wait for you to make them more readable.